apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -0,0 +1,626 @@
+            }
+            // Awaits until batch is sent and response received.
+            IList<string> ids = await _taskHelper.ConfigureAwait(batchTask);
+            // Return the message ID sent from the server.
+            return ids[index];
+        }
+
+        /// <inheritdoc/>
+        public override FlowState GetCurrentFlowState() => _lock.Locked(() => new FlowState(_queueElementCount, _queueByteCount));
+
+        /// <inheritdoc/>
+        public override Task ShutdownAsync(CancellationToken hardStopToken)
+        {
+            lock (_lock)
+            {
+                CheckShutdown();
+                _softStopCts.Cancel();
+                QueueCurrentBatchIfRequired(true, 0);
+                var registration = hardStopToken.Register(() =>
+                {
+                    _hardStopCts.Cancel();
+                    ShutdownIfCompleted();
+                });
+                _taskHelper.Run(async () =>
+                {
+                    await _taskHelper.ConfigureAwaitHideErrors(_shutdownTcs.Task);
+                    registration.Dispose();
+                });
+                ShutdownIfCompleted();
+            }
+            return _shutdownTcs.Task;
+        }
+
+        private void CheckShutdown() => GaxPreconditions.CheckState(!_softStopCts.IsCancellationRequested, "Publisher already shutdown, cannot use");
+
+        private void ShutdownIfCompleted()
+        {
+            // Pre-condition: Must be locked if not a hard stop
+            if (_hardStopCts.IsCancellationRequested || (_softStopCts.IsCancellationRequested && _batchesInFlightCount == 0 && _batchesReady.Count == 0))
+            {
+                if (_shutdownLock.Locked(() => {
+                    var shutdownStarted = _shutdownStarted;
+                    _shutdownStarted = true;
+                    return !shutdownStarted;
+                }))
+                {
+                    // All batches sent and shutdown requested, so signal shutdown completed successfully.
+                    _taskHelper.Run(async () =>
+                    {
+                        if (_shutdown != null)
+                        {
+                            await _taskHelper.ConfigureAwaitHideErrors(_shutdown());
+                        }
+                        if (_hardStopCts.IsCancellationRequested)
+                        {
+                            _shutdownTcs.SetCanceled();
+                        }
+                        else
+                        {
+                            _shutdownTcs.SetResult(0);
+                        }
+                    });
+                }
+            }
+        }
+
+        // Not "async void" to gain standard async exeption handling
+        private async Task CreateBatch()
+        {
+            // Pre-condition: Must be locked - lock active up until first 'await'.
+            _currentBatch = new Batch();
+            var timerToken = _currentBatch.TimerCts.Token;
```

Comment:
Not the cancellation token part really - more that there's an async method which is "create batch and run it later". Possibly bad choice on my part of where to put the comment.